tests/partitions_test.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+
+import infra.network
+import infra.e2e_args
+import infra.partitions
+import infra.logging_app as app
+import suite.test_requirements as reqs
+
+from loguru import logger as LOG
+
+
+@reqs.description("Invalid partitions are not allowed")
+def test_invalid_partitions(network, args):
+    nodes = network.get_joined_nodes()
+
+    try:
+        network.partitioner.partition(
+            [nodes[0], nodes[2]],
+            [nodes[1], nodes[2]],
+        )
+        assert False, "Node should not appear in two or more partitions"
+    except ValueError:
+        pass
+
+    try:
+        network.partitioner.partition()
+        assert False, "At least one partition should be specified"
+    except ValueError:
+        pass
+
+    try:
+        invalid_local_node_id = -1
+        new_node = infra.node.Node(invalid_local_node_id, "local://localhost")
+        network.partitioner.partition([new_node])
+        assert False, "All nodes should belong to network"
+    except ValueError:
+        pass
+
+    return network
+
+
+@reqs.description("Partition primary + f nodes")
+def test_partition_majority(network, args):
+    primary, backups = network.find_nodes()
+
+    # Create a partition with primary + half remaining nodes (i.e. majority)
+    partition = [backups[0]]
+    partition.extend(backups[len(backups) // 2 :])
+
+    # Note: Context manager
+    with network.partitioner.partition(partition):
+        try:
+            network.wait_for_new_primary(primary.node_id)
+            assert False, "No new primary should be elected when partitioning majority"
+        except TimeoutError:
+            pass
+
+    return network
```

Comment:
Minor: Although there's no election during this partition, there will be one when the network anneals. Any node that has been partitioned from the primary will start calling elections, and when it rejoins the network it will continue to call them. It can't win if more transactions have come in (someone in the majority partition that has the up-to-date state will win*), but it may still cause a leader change. This might be a problem for whichever test starts next, so I think we should wait for all nodes to settle down after removing partitions.

\* This is probably worth testing in this test. While the primary is in a majority partition, we can actually submit transaction to the primary or any of the in-partition backups, and these will commit and persist.